File: data/solver/src/cs/Cafe.cs
using System;
using System.Collections.Generic;

namespace CoffeeDistribution.Entities
{
    public class Cafe
    {
        public string Name { get; private set; }
        public string Location { get; private set; }
        public string Contact { get; private set; }
        public Dictionary<string, int> CoffeeDemand { get; private set; }

        public Cafe(string name, string location, string contact)
        {
            Name = name;
            Location = location;
            Contact = contact;
            CoffeeDemand = new Dictionary<string, int>();
        }

        public void SetCoffeeDemand(string coffeeType, int quantity)
        {
            CoffeeDemand[coffeeType] = quantity;
        }

        public int GetCoffeeDemand(string coffeeType)
        {
            return CoffeeDemand.TryGetValue(coffeeType, out int quantity) ? quantity : 0;
        }

        public string DisplayInfo()
        {
            var info = $"Name: {Name}\nLocation: {Location}\nContact: {Contact}\nCoffee Demand:\n";
            foreach (var demand in CoffeeDemand)
            {
                info += $"  - {demand.Key}: {demand.Value}\n";
            }
            return info;
        }

        public override string ToString()
        {
            return $"Cafe(name={Name}, location={Location}, contact={Contact})";
        }

        public void FulfillDemand(string coffeeType, int quantity)
        {
            int currentDemand = GetCoffeeDemand(coffeeType);
            if (quantity > currentDemand)
            {
                throw new ArgumentException("Fulfilled quantity exceeds demand");
            }
            CoffeeDemand[coffeeType] = currentDemand - quantity;
        }
    }
}

File: data/solver/src/cpp/Cafe.cpp
#include "Cafe.h"
#include <iostream>
#include <stdexcept>

void Cafe::display_info() const {
    std::cout << "Name: " << name << "\nLocation: " << location << "\nContact: " << contact << "\nCoffee Demand:\n";
    for (const auto& demand : coffee_demand) {
        std::cout << "  - " << demand.first << ": " << demand.second << "\n";
    }
}

std::string Cafe::to_string() const {
    return "Cafe(name=" + name + ", location=" + location + ", contact=" + contact + ")";
}

void Cafe::fulfill_demand(const std::string& coffee_type, int quantity) {
    int current_demand = get_coffee_demand(coffee_type);
    if (quantity > current_demand) {
        throw std::runtime_error("Fulfilled quantity exceeds demand");
    }
    coffee_demand[coffee_type] = current_demand - quantity;
}